src/Servers/Kestrel/Core/src/Internal/Http2/Http2StreamOfT.cs

```diff
@@ -9,22 +13,81 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2
 {
     internal sealed class Http2Stream<TContext> : Http2Stream, IHostContextContainer<TContext> where TContext : notnull
     {
-        private readonly IHttpApplication<TContext> _application;
+        private readonly Task _executingTask;
+        private bool _disposed;
 
-        public Http2Stream(IHttpApplication<TContext> application, Http2StreamContext context) 
+        private readonly Channel<ReadResult> _requestQueue = Channel.CreateUnbounded<ReadResult>(new UnboundedChannelOptions
```

Comment:
This could be bounded to a single element, right?

Reply:
Also, if we try this for real, we should probably move this to the base type and use a private class rather than abuse ReadResult.

Reply:
Errr which base type?

Reply:
BoundedChannels are bigger (they store more state since both the readers and writers can block). I could actually try to use the ReusableTcs that @JamesNK wrote for the other loop. It stores much less state and is lighter.